aten/src/ATen/native/hammerblade/Offload.h

```diff
@@ -1,102 +1,138 @@
+//==============================================
+
+//----------------------------------------------
+// 0-3 tensors + 0 scalar
+//----------------------------------------------
+
+inline void hb_offload_kernel(const char* kernel) {
+  std::vector<Tensor> args;
+  std::vector<Scalar> scalars;
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  std::vector<Scalar> scalars;
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  std::vector<Scalar> scalars;
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, Tensor t2, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  args.push_back(t2);
+  std::vector<Scalar> scalars;
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, Tensor t2, Tensor t3,
+                              const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  args.push_back(t2);
+  args.push_back(t3);
+  std::vector<Scalar> scalars;
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+//----------------------------------------------
+// 0-3 tensors + 1 scalar
+//----------------------------------------------
+
+inline void hb_offload_kernel(Scalar s0, const char* kernel) {
+  std::vector<Tensor> args;
+  std::vector<Scalar> scalars;
+  scalars.push_back(s0);
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Scalar s0, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  std::vector<Scalar> scalars;
+  scalars.push_back(s0);
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, Scalar s0, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  std::vector<Scalar> scalars;
+  scalars.push_back(s0);
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, Tensor t2, Scalar s0,
+                              const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  args.push_back(t2);
+  std::vector<Scalar> scalars;
+  scalars.push_back(s0);
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
+
+inline void hb_offload_kernel(Tensor t0, Tensor t1, Tensor t2, Tensor t3,
+                              Scalar s0, const char* kernel) {
+  std::vector<Tensor> args;
+  args.push_back(t0);
+  args.push_back(t1);
+  args.push_back(t2);
+  args.push_back(t3);
+  std::vector<Scalar> scalars;
+  scalars.push_back(s0);
+  offload_tensor_scalar_impl(args, scalars, kernel);
+}
```

Comment:
This looks great! 

I spent a day or so looking for variadic templates to make this more elegant... this needs "compile-time looping". I found c++11 doesn't have much support for that.

Reply:
yeah that would super hard I guess